### DISCUSSION TITLE
How do I reference the "nuget.exe" in downloads/tools from CMAKE

### DISCUSSION THREAD

**Discussion by ag-ctlg** [microsoft/vcpkg#39799] [Q&A] [answered]
Hi everyone,
I'm trying to port some C++ projects from VS/MSBuild to CMAKE+vcpkg. The projects have some NuGet package dependencies and I've created a CMake script to use a "nuget.exe" to restore packages listed in a packages.config file. I'm using CMakePresets.json and CMakeUserPresets.json and have created a CMake variable to set the path to the "nuget.exe".

I just saw that vcpkg installs its own nuget.exe under "downloads/tools/nuget_VERSION/" and I would like to use this instead of relying on developers (and the build server) having a nuget.exe installed localy and adapting the CMakeUserPresets.json to point to the installed folder.

How do I reference the vcpkg installed nuget.exe from my CMake scripts? I thought about using "$env{VCPKG_DOWNLOADS}/tools/.../nuget.exe" but that would require me to hardcode the version in the path (e.g. /tools/nuget-6.10.0-windows). I would just like to have a reference to whatever the installed nuget.exe is. Is this possible or would it require changes to vpckg, i.e. adding a new environment variable or some other plumbing?

Or I should just use CMake globbing functions to get a path to the nuget.exe? 🤨

Thank you for your help ☺️

**Comment by AugP** [comment, marked as answer]
Posting a comment from @BillyONeal: You can try running `vcpkg fetch nuget` to get the copy of nuget.exe vcpkg uses, but that is an internal and undocumented vcpkg command and it could be subject to breaking changes in the future if we ever need to modify its behavior.

**Reply by kwsp** [reply]
@AugP You say `vcpkg fetch nuget` is an internal and undocumented command but many official tutorials refer to it. I guess it's more public and used than you may realise

* https://learn.microsoft.com/en-us/vcpkg/consume/binary-caching-nuget
* https://learn.microsoft.com/en-us/vcpkg/consume/binary-caching-github-packages

**Reply by ag-ctlg** [reply]
Great, thank you guys! So I can then do this:
`execute_process(COMMAND "$ENV{VCPKG_ROOT}/vcpkg.exe" fetch nuget OUTPUT_VARIABLE nuget_exe)`
and then use the `execute_process(COMMAND ${nuget_exe} restore ...) `

👍

**Reply by kwsp** [reply]
Careful the output of `vcpkg fetch nuget` sometimes prints a lot of junk, but I think the last line is guaranteed to be the path to the executable, hence in the bash examples you see it used like `vcpkg fetch nuget | tail -n 1`

**Reply by ag-ctlg** [reply]
There is an issue though that I don't really understand why it's happenning. When I use the `vcpkg fetch nuget` I get a full path and store it in a cmake variable like this:
`execute_process(COMMAND "$ENV{VCPKG_ROOT}/vcpkg.exe" fetch nuget OUTPUT_VARIABLE _VCPKG_NUGET_PATH)` 

Then later on if I use `execute_process(COMMAND "${_VCPKG_NUGET_PATH}" restore ...)` it spits out an error (but only if `COMMAND_ERROR_IS_FATAL` is set) about file not existing. However, if I use `cmake_path(GET _VCPKG_NUGET_PATH PARENT_PATH NUGET_EXE_FOLDER)` and then `execute_process(COMMAND "${NUGET_EXE_FOLDER}/nuget.exe" restore ...)` then it works. And the path looks the same as `_VCPKG_NUGET_PATH`.
